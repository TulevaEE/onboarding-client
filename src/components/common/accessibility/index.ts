import { useEffect, useState } from 'react';

// synced with wordpress-theme
const HIGH_CONTRAST_MODE_COOKIE_NAME = 'high-contrast';

export const useHighContrastMode = () => {
  const [enabled, setEnabled] = useState(isHighContrastModeEnabled() ?? getInitialState);

  const toggleContrastMode = (switchOn: boolean) => {
    toggleHighContrastMode(switchOn);
    setEnabled(isHighContrastModeEnabled() ?? getInitialState);
  };

  useEffect(() => {
    document.documentElement.classList.toggle('high-contrast', enabled ?? undefined);
  }, [enabled]);

  return { enabled, toggleContrastMode };
};

const toggleHighContrastMode = (switchOn: boolean) => {
  const domain = window.location.host.includes('localhost') ? 'localhost' : '.tuleva.ee';

  const oneDayInSeconds = 60 * 60 * 24;
  const cookieLifetimeInSeconds = oneDayInSeconds * 400;
  document.cookie = `${HIGH_CONTRAST_MODE_COOKIE_NAME}=${switchOn};max-age=${cookieLifetimeInSeconds};domain=${domain};path=/`;
};

const getInitialState = () => window.matchMedia('(prefers-contrast: more)').matches;

const isHighContrastModeEnabled = () => {
  const match = document.cookie.match(
    new RegExp(`(?:^|;)\\s*${HIGH_CONTRAST_MODE_COOKIE_NAME}=([^;]*)`),
  );

  if (!match) {
    return null;
  }

  return match[1] === 'true';
};
