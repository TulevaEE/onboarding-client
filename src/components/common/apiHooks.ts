import {
  useMutation,
  UseMutationResult,
  useQuery,
  useQueryClient,
  UseQueryResult,
} from '@tanstack/react-query';

import {
  contactMemberCapitalListingOwner,
  createApplicationCancellation,
  createCapitalTransferContract,
  createMandateBatch,
  createMemberCapitalListing,
  createSavingsFundPaymentCancellation,
  deleteMemberCapitalListing,
  getCapitalEvents,
  getCapitalRowsWithToken,
  getCapitalTotal,
  getCapitalTransferContract,
  getContributions,
  getFundPensionStatus,
  getFunds,
  getMandateDeadlines,
  getMemberCapitalListingCount,
  getMemberCapitalListings,
  getMyCapitalTransferContracts,
  getPendingApplications,
  getSavingsFundBalance,
  getSavingsFundOnboardingStatus,
  getSourceFunds,
  getTransactions,
  getUserConversionWithToken,
  getUserWithToken,
  getWithdrawalsEligibility,
  previewMessageForMemberCapitalListing,
  updateCapitalTransferContract,
} from './api';
import {
  Application,
  CancellationMandate,
  CapitalEvent,
  CapitalRow,
  CapitalTotal,
  ContactListingOwnerDto,
  ContactListingOwnerResponse,
  Contribution,
  CreateMemberCapitalListingDto,
  ErrorResponse,
  Fund,
  MandateDeadlines,
  MemberCapitalListing,
  SavingsFundOnboardingStatus,
  SavingsFundPaymentCancellationCommand,
  SourceFund,
  Transaction,
  User,
  UserConversion,
} from './apiModels';
import {
  CreateMandateBatchDto,
  FundPensionStatus,
  MandateBatchDto,
  WithdrawalsEligibility,
} from './apiModels/withdrawals';
import {
  CapitalTransferContract,
  CreateCapitalTransferDto,
  UpdateCapitalTransferContractDto,
} from './apiModels/capital-transfer';

export function usePendingApplications(): UseQueryResult<Application[]> {
  return useQuery(['pendingApplications'], () => getPendingApplications());
}

export function useTransactions(): UseQueryResult<Transaction[]> {
  return useQuery(['transactions'], () => getTransactions());
}

export function useCapitalEvents(): UseQueryResult<CapitalEvent[]> {
  return useQuery(['capitalEvents'], () => getCapitalEvents());
}

export function useCapitalRows(): UseQueryResult<CapitalRow[]> {
  return useQuery(['capitalRows'], () => getCapitalRowsWithToken());
}

export function useCapitalTotal(): UseQueryResult<CapitalTotal> {
  return useQuery(['capitalTotal'], () => getCapitalTotal());
}

export function useContributions(): UseQueryResult<Contribution[]> {
  return useQuery(['contributions'], () => getContributions());
}

export function useFunds(): UseQueryResult<Fund[]> {
  return useQuery(['funds'], () => getFunds());
}

export function useApplicationCancellation(): UseMutationResult<
  CancellationMandate,
  unknown,
  number,
  unknown
> {
  return useMutation((applicationId: number) => createApplicationCancellation(applicationId));
}

export function useSavingsFundPaymentCancellation(): UseMutationResult<
  void,
  unknown,
  SavingsFundPaymentCancellationCommand,
  unknown
> {
  return useMutation(({ paymentId }: SavingsFundPaymentCancellationCommand) =>
    createSavingsFundPaymentCancellation(paymentId),
  );
}

export function useApplication(id: number): UseQueryResult<Application | null> {
  // TODO: replace with /:id api once one exists
  const result = usePendingApplications();
  return {
    ...result,
    data: result.data ? result.data.find((application) => application.id === id) || null : null,
  } as UseQueryResult<Application | null>;
}

export function useMandateDeadlines(): UseQueryResult<MandateDeadlines> {
  return useQuery(['mandateDeadlines'], () => getMandateDeadlines());
}

export function useMe(): UseQueryResult<User> {
  return useQuery(['user'], () => getUserWithToken());
}

export function useWithdrawalsEligibility(): UseQueryResult<WithdrawalsEligibility> {
  return useQuery(['withdrawalsEligibility'], () => getWithdrawalsEligibility());
}

export function useFundPensionStatus(): UseQueryResult<FundPensionStatus> {
  return useQuery(['fundPensionStatus'], () => getFundPensionStatus());
}

export function useConversion(): UseQueryResult<UserConversion> {
  return useQuery(['conversion'], () => getUserConversionWithToken());
}

export function useSourceFunds(fromDate?: string, toDate?: string): UseQueryResult<SourceFund[]> {
  return useQuery(['sourceFunds', fromDate, toDate], () => getSourceFunds(fromDate, toDate));
}

export function useSavingsFundOnboardingStatus(): UseQueryResult<SavingsFundOnboardingStatus> {
  return useQuery(['savingsFundOnboardingStatus'], () => getSavingsFundOnboardingStatus());
}

export function useCreateMandateBatch(): UseMutationResult<
  MandateBatchDto,
  ErrorResponse,
  CreateMandateBatchDto,
  unknown
> {
  return useMutation((dto) => createMandateBatch(dto));
}

export function useCreateCapitalTransferContract(): UseMutationResult<
  CapitalTransferContract,
  ErrorResponse,
  CreateCapitalTransferDto,
  unknown
> {
  return useMutation((dto) => createCapitalTransferContract(dto));
}

export function useMyCapitalTransferContracts(): UseQueryResult<CapitalTransferContract[]> {
  return useQuery(['myCapitalTransferContracts'], () => getMyCapitalTransferContracts());
}

export function useCapitalTransferContract(
  id: number,
  manualRefetch: boolean,
): UseQueryResult<CapitalTransferContract, { status: number }> {
  return useQuery(['capitalTransferContract', id], () => getCapitalTransferContract(id), {
    enabled: !manualRefetch,
    retry: false,
  });
}

export function useUpdateCapitalTransferContract(): UseMutationResult<
  CapitalTransferContract,
  ErrorResponse,
  UpdateCapitalTransferContractDto,
  unknown
> {
  const queryClient = useQueryClient();

  return useMutation((dto) => updateCapitalTransferContract(dto), {
    onSuccess: () => {
      queryClient.invalidateQueries(['myCapitalTransferContracts']);
    },
  });
}

export function useMemberCapitalListings(): UseQueryResult<MemberCapitalListing[]> {
  return useQuery(['memberCapitalListings'], () => getMemberCapitalListings());
}

export function useMemberCapitalListingCount(): UseQueryResult<number> {
  return useQuery(['memberCapitalListingCount'], () => getMemberCapitalListingCount());
}

export function useCreateMemberCapitalListing(): UseMutationResult<
  MemberCapitalListing,
  ErrorResponse,
  CreateMemberCapitalListingDto,
  unknown
> {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (dto) => createMemberCapitalListing(dto),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['memberCapitalListings'] });
    },
  });
}

export function useDeleteMemberCapitalListing(): UseMutationResult<
  MemberCapitalListing,
  ErrorResponse,
  MemberCapitalListing,
  unknown
> {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (listing) => deleteMemberCapitalListing(listing),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['memberCapitalListings'] });
    },
  });
}

export function usePreviewMemberCapitalListingMessage(dto: ContactListingOwnerDto) {
  return useQuery(
    ['listingMessage', dto.id, dto.addPersonalCode, dto.addPhoneNumber],
    () => previewMessageForMemberCapitalListing(dto),
    {
      enabled: true,
    },
  );
}

export function useContactMemberCapitalListing(): UseMutationResult<
  ContactListingOwnerResponse,
  ErrorResponse,
  ContactListingOwnerDto,
  unknown
> {
  return useMutation({
    mutationFn: (dto) => contactMemberCapitalListingOwner(dto),
  });
}

export function useSavingsFundBalance(): UseQueryResult<SourceFund | null> {
  return useQuery(['savingsFundBalance'], () => getSavingsFundBalance());
}
