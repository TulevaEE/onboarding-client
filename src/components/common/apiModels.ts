export type Application =
  | TransferApplication
  | StopContributionsApplication
  | ResumeContributionsApplication
  | EarlyWithdrawalApplication
  | WithdrawalApplication
  | PaymentApplication;

export type PaymentApplication = BaseApplication<
  ApplicationType.PAYMENT,
  {
    amount: number;
    currency: Currency;
    targetFund: Fund;
  }
>;

export type TransferApplication = BaseApplication<
  ApplicationType.TRANSFER,
  {
    sourceFund: Fund;
    exchanges: {
      targetFund: Fund | null;
      targetPik: string | null;
      amount: number;
    }[];
    cancellationDeadline: string;
  }
>;

export type StopContributionsApplication = BaseApplication<
  ApplicationType.STOP_CONTRIBUTIONS,
  {
    stopTime: string;
    earliestResumeTime: string;
    cancellationDeadline: string;
  }
>;

export type ResumeContributionsApplication = BaseApplication<
  ApplicationType.RESUME_CONTRIBUTIONS,
  {
    resumeTime: string;
    cancellationDeadline: string;
  }
>;

export type EarlyWithdrawalApplication = BaseApplication<
  ApplicationType.EARLY_WITHDRAWAL,
  {
    depositAccountIBAN: string;
    cancellationDeadline: string;
    fulfillmentDate: string;
  }
>;

export type WithdrawalApplication = BaseApplication<
  ApplicationType.WITHDRAWAL,
  {
    depositAccountIBAN: string;
    cancellationDeadline: string;
    fulfillmentDate: string;
  }
>;

export enum ApplicationType {
  TRANSFER = 'TRANSFER',
  STOP_CONTRIBUTIONS = 'STOP_CONTRIBUTIONS',
  RESUME_CONTRIBUTIONS = 'RESUME_CONTRIBUTIONS',
  EARLY_WITHDRAWAL = 'EARLY_WITHDRAWAL',
  WITHDRAWAL = 'WITHDRAWAL',
  PAYMENT = 'PAYMENT',
}

export enum ApplicationStatus {
  PENDING = 'PENDING',
  COMPLETE = 'COMPLETE',
  FAILED = 'FAILED',
}

export interface BaseApplication<Type extends ApplicationType, Details> {
  id: number;
  status: ApplicationStatus;
  creationTime: string;

  type: Type;
  details: Details;
}

export interface Fund {
  isin: string;
  name: string;
  pillar: Pillar;
  managementFeeRate: number;
  ongoingChargesFigure: number;
  fundManager: FundManager;
  status: FundStatus;
}

export type Pillar = 2 | 3;

enum FundStatus {
  ACTIVE = 'ACTIVE',
  LIQUIDATED = 'LIQUIDATED',
  SUSPENDED = 'SUSPENDED',
  CONTRIBUTIONS_FORBIDDEN = 'CONTRIBUTIONS_FORBIDDEN',
  PAYOUTS_FORBIDDEN = 'PAYOUTS_FORBIDDEN',
}

export interface FundBalance {
  fund: Fund;
  value: number;
  unavailableValue: number;
  currency: string;
  activeContributions: boolean;
  contributions: number;
  subtractions: number;
  profit: number;
}

export interface SourceFund {
  fundManager: FundManager;
  activeFund: boolean;
  name: string;
  pillar: Pillar;
  managementFeePercent: string;
  isin: string;
  price: number;
  unavailablePrice: number;
  currency: string;
  ongoingChargesFigure: number;
  contributions: number;
  subtractions: number;
  profit: number;
}

export interface FundManager {
  name: string;
}

export interface CancellationMandate {
  mandateId: number;
}

export interface Mandate {
  id: number;
}

interface Address {
  street: string;
  districtCode: string;
  postalCode: string;
  countryCode: string;
}

export interface User {
  id: number;
  personalCode: string;
  firstName: string;
  lastName: string;
  email: string;
  phoneNumber: string;
  memberNumber: number;
  pensionAccountNumber: string;
  secondPillarPikNumber: string;
  address: Address;
  secondPillarActive: boolean;
  thirdPillarActive: boolean;
  age: number;
}

export interface UserConversion {
  secondPillar: Conversion;
  thirdPillar: Conversion;
}

export interface Amount {
  yearToDate: number;
  total: number;
}

export interface Conversion {
  selectionComplete: boolean;
  transfersComplete: boolean;
  paymentComplete: boolean;
  pendingWithdrawal: boolean;
  contribution: Amount;
  subtraction: Amount;
}

export interface InitialCapital {
  membershipBonus: number;
  capitalPayment: number;
  currency: string;
  unvestedWorkCompensation: number;
  workCompensation: number;
  profit: number;
}

export interface AmlCheck {
  type: string;
  success: boolean;
}

export interface HttpError {
  body: { errors: [{ code: string }] };
}

export interface Token {
  accessToken: string;
}

export interface MobileSignatureResponse {
  challengeCode: string;
}

export interface MobileSignatureStatusResponse {
  statusCode: string;
  challengeCode: string;
}

export interface IdCardSignatureResponse {
  hash: string;
}

export interface IdCardSignatureStatusResponse {
  statusCode: string;
}

export interface Payment {
  amount: number;
  currency: Currency;
  bank: string;
}

export type Currency = 'EUR';
