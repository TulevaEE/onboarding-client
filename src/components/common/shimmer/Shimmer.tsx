import React from 'react';
import styles from './Shimmer.module.scss';

export const Shimmer: React.FunctionComponent<{ height: number }> = ({ height }) => {
  return <div className={styles.shimmer} style={{ height: `${height}px` }} />;
};
