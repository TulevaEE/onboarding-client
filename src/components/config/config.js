import config from 'react-global-configuration';

const BASE_CONFIG = {
  clientCredentialsAccessToken: undefined,
  language: 'et',
  idCardUrl: 'https://id.tuleva.ee',
  memberFee: 125,
};

const ENV_CONFIGS = {
  development: {
    ...BASE_CONFIG,
    applicationUrl: 'http://localhost:3000',
    clientCredentialsAccessToken: '6b338ba2-805c-4300-9341-b38bb4ad34a9',
  },
  production: {
    ...BASE_CONFIG,
    applicationUrl: 'https://pension.tuleva.ee',
    clientCredentialsAccessToken: '705e26c1-9316-47f2-94b8-a5c6b0dfb566',
  },
  staging: {
    ...BASE_CONFIG,
    applicationUrl: 'https://staging.tuleva.ee',
    clientCredentialsAccessToken: '6b338ba2-805c-4300-9341-b38bb4ad34a9',
    idCardUrl: 'https://id-staging.tuleva.ee',
  },
  test: {
    ...BASE_CONFIG,
    applicationUrl: 'http://localhost',
    language: 'en',
    clientCredentialsAccessToken: undefined,
  },
};

export function initializeConfiguration() {
  const env = process.env.NODE_ENV || process.env.REACT_APP_ENV;
  const selectedConfig = ENV_CONFIGS[env] || BASE_CONFIG;
  config.set(selectedConfig, { freeze: false });
}

export function updateLanguage(language) {
  const env = process.env.NODE_ENV || process.env.REACT_APP_ENV;
  const selectedConfig = { ...ENV_CONFIGS[env], language };
  config.set(selectedConfig, { freeze: false });
}
