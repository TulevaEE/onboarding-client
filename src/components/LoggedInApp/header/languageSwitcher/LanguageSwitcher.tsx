import React, { FC } from 'react';
import config from 'react-global-configuration';

const LanguageSwitcher: FC = () => (
  <>
    {config.get('language') === 'et' ? (
      <a className="btn btn-link p-0 border-0" href="?language=en">
        EN
      </a>
    ) : (
      <a className="btn btn-link p-0 border-0" href="?language=et">
        ET
      </a>
    )}
  </>
);

export default LanguageSwitcher;
