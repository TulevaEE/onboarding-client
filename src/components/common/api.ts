import config from 'react-global-configuration';
import {
  AmlCheck,
  Application,
  Authentication,
  CancellationMandate,
  CapitalEvent,
  CapitalRow,
  Contribution,
  Fund,
  FundBalance,
  IdCardSignatureResponse,
  LoginMethod,
  Mandate,
  MandateDeadlines,
  MobileSignatureResponse,
  MobileSignatureStatusResponse,
  Payment,
  PaymentLink,
  PaymentType,
  SourceFund,
  Token,
  Transaction,
  User,
  UserConversion,
} from './apiModels';
import {
  downloadFileWithAuthentication,
  getWithAuthentication,
  patchWithAuthentication,
  post,
  postForm,
  postWithAuthentication,
  putWithAuthentication,
  simpleFetch,
} from './http';
import { getAuthentication } from './authenticationManager';
import {
  CreateMandateBatchDto,
  FundPensionStatus,
  MandateBatchDto,
  WithdrawalsEligibility,
} from './apiModels/withdrawals';

const API_URI = '/api';

export function getEndpoint(endpoint: string): string {
  // in production, we proxy through a proxy endpoint at /proxy.
  // in development, we proxy through webpack dev server without the prefix.
  if (process.env.NODE_ENV === 'production') {
    const API_BASEPATH = config && config.get('applicationUrl') ? config.get('applicationUrl') : '';
    return `${API_BASEPATH}${API_URI}${endpoint}`;
  }
  if (process.env.NODE_ENV === 'test') {
    const API_BASEPATH = config && config.get('applicationUrl') ? config.get('applicationUrl') : '';
    return `${API_BASEPATH}${endpoint}`;
  }
  return endpoint;
}

export async function authenticateWithMobileId(
  phoneNumber: string,
  personalCode: string,
): Promise<string> {
  const { challengeCode } = await post(getEndpoint('/authenticate'), {
    phoneNumber,
    personalCode,
    type: 'MOBILE_ID',
  });
  return challengeCode;
}

export async function authenticateWithIdCode(personalCode: string): Promise<Authentication> {
  const { challengeCode, authenticationHash } = await post(getEndpoint('/authenticate'), {
    personalCode,
    type: 'SMART_ID',
  });
  return { challengeCode, authenticationHash };
}

export async function authenticateWithIdCard(): Promise<boolean> {
  await simpleFetch('GET', config.get('idCardUrl')); // http://stackoverflow.com/a/16818527
  const { success } = await simpleFetch('POST', `${config.get('idCardUrl')}/idLogin`);
  return success;
}

export function logout(): Promise<void> {
  const request = getWithAuthentication(getEndpoint('/v1/logout'), {});
  getAuthentication().remove();
  return request;
}

export async function getTokensWithGrantType(
  grantType: LoginMethod,
  extraParameters: Record<string, string> = {},
): Promise<Token | null> {
  try {
    const { access_token: accessToken, refresh_token: refreshToken } = await postForm(
      getEndpoint('/oauth/token'),
      {
        grant_type: grantType,
        client_id: 'onboarding-client',
        ...extraParameters,
      },
      {
        Authorization: 'Basic b25ib2FyZGluZy1jbGllbnQ6b25ib2FyZGluZy1jbGllbnQ=',
      },
    );

    getAuthentication().update({
      accessToken,
      refreshToken,
      loginMethod: grantType,
    });
    return { accessToken, refreshToken };
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  } catch (error: any) {
    if (error.error !== 'AUTHENTICATION_NOT_COMPLETE') {
      throw error;
    }
    return null;
  }
}

export function getMobileIdTokens(): Promise<Token | null> {
  return getTokensWithGrantType('MOBILE_ID');
}

export function getSmartIdTokens(authenticationHash: string): Promise<Token | null> {
  return getTokensWithGrantType('SMART_ID', { authenticationHash });
}

export function getIdCardTokens(): Promise<Token | null> {
  return getTokensWithGrantType('ID_CARD');
}

export function downloadMandatePreviewWithId(entityId: string): Promise<Blob> {
  return downloadFileWithAuthentication(getEndpoint(`/v1/mandates/${entityId}/file/preview`));
}

export function downloadMandateWithId(entityId: string): Promise<Blob> {
  return downloadFileWithAuthentication(getEndpoint(`/v1/mandates/${entityId}/file`));
}

export function getUserWithToken(): Promise<User> {
  return getWithAuthentication(getEndpoint('/v1/me'), undefined);
}

export function getWithdrawalsEligibility(): Promise<WithdrawalsEligibility> {
  return getWithAuthentication(getEndpoint('/v1/withdrawals/eligibility'), undefined);
}

export function getFundPensionStatus(): Promise<FundPensionStatus> {
  return getWithAuthentication(getEndpoint('/v1/withdrawals/fund-pension-status'), undefined);
}

export function createMandateBatch(
  createMandateBatchDto: CreateMandateBatchDto,
): Promise<MandateBatchDto> {
  return postWithAuthentication(getEndpoint('/v1/mandate-batches'), createMandateBatchDto);
}

export async function getSourceFunds(): Promise<SourceFund[]> {
  const funds = await getWithAuthentication(
    getEndpoint('/v1/pension-account-statement'),
    undefined,
  );
  return funds.map(transformFundBalance);
}

const transformFundBalance = (fundBalance: FundBalance): SourceFund => ({
  isin: fundBalance.fund.isin,
  price: fundBalance.value,
  unavailablePrice: fundBalance.unavailableValue,
  activeFund: fundBalance.activeContributions,
  currency: fundBalance.currency || 'EUR',
  name: fundBalance.fund.name,
  fundManager: fundBalance.fund.fundManager,
  managementFeePercent: +(fundBalance.fund.managementFeeRate * 100).toFixed(2).replace(/0+$/, ''),
  pillar: fundBalance.fund.pillar,
  ongoingChargesFigure: fundBalance.fund.ongoingChargesFigure,
  contributions: fundBalance.contributions,
  subtractions: fundBalance.subtractions,
  profit: fundBalance.profit,
});

export function saveMandateWithAuthentication(mandate: string): Promise<Mandate> {
  return postWithAuthentication(getEndpoint('/v1/mandates'), mandate);
}

export async function getMobileIdSignatureChallengeCode({
  entityId,
  type = 'MANDATE',
}: {
  entityId: string;
  type?: 'MANDATE' | 'MANDATE_BATCH';
}): Promise<string | null> {
  const path =
    type === 'MANDATE'
      ? `/v1/mandates/${entityId}/signature/mobileId`
      : `/v1/mandate-batches/${entityId}/signature/mobile-id`;

  const { challengeCode } = await putWithAuthentication<MobileSignatureResponse>(
    getEndpoint(path),
    undefined,
  );
  return challengeCode;
}

export async function getMobileIdSignatureStatus({
  entityId,
  type = 'MANDATE',
}: {
  entityId: string;
  type?: 'MANDATE' | 'MANDATE_BATCH';
}): Promise<MobileSignatureStatusResponse> {
  const path =
    type === 'MANDATE'
      ? `/v1/mandates/${entityId}/signature/mobileId/status`
      : `/v1/mandate-batches/${entityId}/signature/mobile-id/status`;

  return getWithAuthentication<MobileSignatureStatusResponse>(getEndpoint(path), undefined);
}

export async function getSmartIdSignatureChallengeCode({
  entityId,
  type = 'MANDATE',
}: {
  entityId: string;
  type?: 'MANDATE' | 'MANDATE_BATCH';
}): Promise<string | null> {
  const path =
    type === 'MANDATE'
      ? `/v1/mandates/${entityId}/signature/smartId`
      : `/v1/mandate-batches/${entityId}/signature/smart-id`;

  const { challengeCode } = await putWithAuthentication<MobileSignatureResponse>(
    getEndpoint(path),
    undefined,
  );
  return challengeCode;
}

export async function getSmartIdSignatureStatus({
  entityId,
  type = 'MANDATE',
}: {
  entityId: string;
  type?: 'MANDATE' | 'MANDATE_BATCH';
}): Promise<MobileSignatureStatusResponse> {
  const path =
    type === 'MANDATE'
      ? `/v1/mandates/${entityId}/signature/smartId/status`
      : `/v1/mandate-batches/${entityId}/signature/smart-id/status`;

  return getWithAuthentication<MobileSignatureStatusResponse>(getEndpoint(path), undefined);
}

export async function getIdCardSignatureHash({
  entityId,
  type = 'MANDATE',
  certificateHex,
}: {
  entityId: string;
  certificateHex: string;
  type?: 'MANDATE' | 'MANDATE_BATCH';
}) {
  const path =
    type === 'MANDATE'
      ? `/v1/mandates/${entityId}/signature/idCard`
      : `/v1/mandate-batches/${entityId}/signature/id-card`;

  const { hash } = await putWithAuthentication<IdCardSignatureResponse>(getEndpoint(path), {
    clientCertificate: certificateHex,
  });

  return hash;
}

export async function getIdCardSignatureStatus({
  entityId,
  type = 'MANDATE',
  signedHash,
}: {
  entityId: string;
  type?: 'MANDATE' | 'MANDATE_BATCH';
  signedHash: string;
}): Promise<string> {
  const path =
    type === 'MANDATE'
      ? `/v1/mandates/${entityId}/signature/idCard/status`
      : `/v1/mandate-batches/${entityId}/signature/id-card/status`;

  const { statusCode } = await putWithAuthentication(getEndpoint(path), { signedHash });
  return statusCode;
}

export function updateUserWithToken(user: User): Promise<User> {
  return patchWithAuthentication(getEndpoint('/v1/me'), user);
}

export function getUserConversionWithToken(): Promise<UserConversion> {
  return getWithAuthentication(getEndpoint('/v1/me/conversion'), undefined);
}

export function getCapitalRowsWithToken(): Promise<CapitalRow[]> {
  return getWithAuthentication(getEndpoint('/v1/me/capital'), undefined);
}

export function createAmlCheck(
  type: string,
  success: boolean,
  metadata: Record<string, unknown>,
): Promise<AmlCheck> {
  return postWithAuthentication(getEndpoint('/v1/amlchecks'), {
    type,
    success,
    metadata,
  });
}

export function getMissingAmlChecks(): Promise<AmlCheck[]> {
  return getWithAuthentication(getEndpoint('/v1/amlchecks'), undefined);
}

export function getFunds(): Promise<Fund[]> {
  return getWithAuthentication(getEndpoint('/v1/funds'));
}

export function getPendingApplications(): Promise<Application[]> {
  return getWithAuthentication(getEndpoint('/v1/applications'), {
    status: 'PENDING',
  });
}

export function getTransactions(): Promise<Transaction[]> {
  return getWithAuthentication(getEndpoint('/v1/transactions'), undefined);
}

export function getCapitalEvents(): Promise<CapitalEvent[]> {
  return getWithAuthentication(getEndpoint('/v1/me/capital/events'), undefined);
}

export function getContributions(): Promise<Contribution[]> {
  return getWithAuthentication(getEndpoint('/v1/contributions'), undefined);
}

export function getMandateDeadlines(): Promise<MandateDeadlines> {
  return getWithAuthentication(getEndpoint('/v1/mandate-deadlines'), undefined);
}

export function createApplicationCancellation(applicationId: number): Promise<CancellationMandate> {
  return postWithAuthentication(getEndpoint(`/v1/applications/${applicationId}/cancellations`), {});
}

export function createTrackedEvent(type: string, data: Record<string, unknown>): Promise<unknown> {
  return postWithAuthentication(getEndpoint('/v1/t'), { type, data });
}

export function getPaymentLink(payment: Payment): Promise<PaymentLink> {
  return getWithAuthentication(getEndpoint('/v1/payments/link'), payment);
}

export async function redirectToPayment(payment: Payment): Promise<void> {
  const wndw = getWindow(payment.type);
  const paymentLink = await getPaymentLink(payment);

  // eslint-disable-next-line no-console
  console.log(
    'Redirecting to:',
    paymentLink.url,
    'Using window:',
    wndw === window ? 'same' : 'new',
    wndw,
  );

  if (isSecurityRestricted(wndw)) {
    // eslint-disable-next-line no-console
    console.warn('Window is restricted, falling back to same window.');
    window.location.replace(paymentLink.url);
    return;
  }

  try {
    wndw.location.replace(paymentLink.url);
  } catch (err) {
    // eslint-disable-next-line no-console
    console.warn('SecurityError detected, falling back to same window:', err);
    window.location.replace(paymentLink.url);
  }
}

function getWindow(paymentType: PaymentType): Window {
  if (paymentType !== 'RECURRING') {
    return window;
  }

  try {
    const newWindow = window.open('', '_blank'); // might be blocked by popup blockers

    if (newWindow == null) {
      // eslint-disable-next-line no-console
      console.warn('New window blocked, falling back to same window');
      return window;
    }

    newWindow.document.write('Loading...');
    return newWindow;
  } catch (error) {
    // eslint-disable-next-line no-console
    console.warn('Error opening new window:', error);
    return window;
  }
}

function isSecurityRestricted(wndw: Window): boolean {
  if (!wndw || wndw.closed) {
    console.warn('Window is null or closed.');
    return true;
  }

  try {
    if (
      typeof wndw.location.href === 'undefined' ||
      wndw.location.href === '' ||
      (wndw.opener && wndw.opener !== window && !wndw.opener.location) ||
      wndw.top !== wndw.self
    ) {
      console.warn('Security restriction detected.');
      return true;
    }
  } catch {
    console.warn('Security error detected while accessing window properties.');
    return true;
  }

  return false;
}
