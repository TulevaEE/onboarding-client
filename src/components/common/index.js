import * as httpModule from './http';
import * as apiModule from './api';
import * as utilsModule from './utils';
import logoModule from './logo.svg';
import LoaderModule from './loader';
import RadioModule from './radio';
import AuthenticationLoaderModule from './authenticationLoader';
import ErrorMessageModule from './errorMessage';
import InfoTooltipModule from './infoTooltip';
import ErrorAlertModule from './errorAlert';

export const http = httpModule;
export const utils = utilsModule;
export const Loader = LoaderModule;
export const Radio = RadioModule;
export const api = apiModule;
export const logo = logoModule;
export const AuthenticationLoader = AuthenticationLoaderModule;
export const ErrorMessage = ErrorMessageModule;
export const InfoTooltip = InfoTooltipModule;
export const ErrorAlert = ErrorAlertModule;
