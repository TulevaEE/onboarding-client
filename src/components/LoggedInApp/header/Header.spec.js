import React from 'react';
import { shallow } from 'enzyme';
import config from 'react-global-configuration';

import { logo } from '../../common';
import { Header } from '.';
import LanguageSwitcher from './languageSwitcher';
import { Shimmer } from '../../common/shimmer/Shimmer';

describe('Header', () => {
  let component;

  beforeAll(() => {
    config.set({ language: 'et' }, { freeze: false });
  });

  beforeEach(() => {
    component = shallow(<Header />);
  });

  it('shows a header logo with a link', () => {
    expect(
      component.contains(
        <a href="/account">
          <img src={logo} alt="Tuleva" className="brand-logo" />
        </a>,
      ),
    ).toBe(true);
  });

  it("shows the user's name when not loading", () => {
    const name = 'A name of a person';
    component.setProps({
      user: { name },
      loading: false,
    });
    expect(component.text()).toContain(name);
    expect(component.contains(<Shimmer height={24} />)).toBe(false);
  });

  it('can log out the user', () => {
    const onLogout = jest.fn();
    component.setProps({ loading: false, onLogout, user: { name: 'name' } });
    expect(onLogout).not.toHaveBeenCalled();
    component.find('[href="/login"]').simulate('click');
    expect(onLogout).toHaveBeenCalledTimes(1);
  });

  it("shows a loader when it's loading", () => {
    component.setProps({ loading: true });
    expect(component.contains(<Shimmer height={24} />)).toBe(true);
  });

  it('renders my account button', () => {
    const onLogout = jest.fn();
    component.setProps({ loading: false, onLogout, user: { name: 'name' } });
    expect(component.find('[href="/account"]').at(1)).toMatchSnapshot();
  });

  it('renders the language switcher', () => {
    component.setProps({ loading: false, user: { name: 'name' } });
    expect(component.contains(<LanguageSwitcher />)).toBe(true);
  });
});
