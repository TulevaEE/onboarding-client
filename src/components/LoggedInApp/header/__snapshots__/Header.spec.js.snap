// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders my account button 1`] = `
<a
  className="icon-link"
  href="/account"
>
  <MemoizedFormattedMessage
    id="header.my.account"
  />
</a>
`;
